testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,297 @@
+ *		CONFIG_COREDUMP (kernel)
+ * Currently it is assumed that if the core dump can not be found it is because
+ * they are disabled rather than a bug. Also tests may pass because of the core
+ * and file size limits.
+ */
+
+static void setup(void)
+{
+	const char *dumpf = "dump-%p";
+	char cwd[1024];
+	char tmpcpattern[1048];
+
+	tc.fd = 0;
+	tc.fname = "mmf0";
+	tc.fmem = NULL;
+	tc.cpsize = -1;
+	tc.cpfd = 0;
+	tc.dfd = 0;
+	tc.ycount = 1280;
+	tc.fmemsize = tc.ycount + 2;
+
+	SAFE_GETCWD(cwd, 1024);
+	snprintf(tmpcpattern, 1048, "%s/%s", cwd, dumpf);
+	tst_res(TINFO, "Temporary core pattern is '%s'", tmpcpattern);
+
+	tc.cpfd = SAFE_OPEN(CORE_PATTERN, O_RDONLY);
+	tc.cpsize = SAFE_READ(0, tc.cpfd, &tc.cpattern, 1048);
+	SAFE_CLOSE(tc.cpfd);
+	tc.cpfd = 0;
+
+	if (tc.cpsize == 1048)
+		tst_brk(TCONF, "Current core pattern is too large to save");
+
+	tc.cpattern[tc.cpsize + 1] = '\0';
+	tst_res(TINFO, "System core pattern is '%s'", tc.cpattern);
+
+	tc.cpfd = SAFE_OPEN(CORE_PATTERN, O_TRUNC | O_WRONLY);
+	SAFE_WRITE(1, tc.cpfd, tmpcpattern, strlen(tmpcpattern));
+	SAFE_CLOSE(tc.cpfd);
+	tc.cpfd = 0;
+}
+
+static void restore_cpattern(void)
+{
+	int fd, rval;
+	size_t written;
+
+	if (tc.cpfd != 0 && close(tc.cpfd))
+		tst_res(TINFO | TERRNO, "close(tc.cpdf)");
+
+	if (tc.cpsize < 1)
+		return;
+
+	fd = open(CORE_PATTERN, O_TRUNC | O_WRONLY);
+	if (fd == -1) {
+		tst_res(TINFO | TERRNO, "open(CORE_PATTERN, O_TRUNC)");
+		goto error;
+	}
+
+	written = write(fd, tc.cpattern, tc.cpsize);
+	if (written < (size_t)tc.cpsize) {
+		tst_brk(TINFO | TERRNO, "write(fd, tc.pattern, ...)");
+		goto error;
+	}
+
+	rval = close(fd);
+	if (rval == -1) {
+		tst_brk(TINFO | TERRNO, "close(fd)");
+		goto error;
+	}
+
+	tc.cpsize = 0;
+	return;
+error:
+	tst_brk(TBROK,
+		"COULD NOT RESTORE "CORE_PATTERN"! It used to be '%s'",
+		tc.cpattern);
+}
+
+static void cleanup(void)
+{
+	restore_cpattern();
+
+	if (tc.fmem != NULL && munmap(tc.fmem, tc.fmemsize))
+		tst_brk(TBROK | TERRNO, "munmap(tc.fmem)");
+	tc.fmem = NULL;
+
+	if (tc.fd > 0 && close(tc.fd))
+		tst_brk(TBROK | TERRNO, "close(tc.fd)");
+
+	if (tc.dfd > 0 && close(tc.dfd))
+		tst_brk(TBROK | TERRNO, "close(tc.dfd)");
+}
+
+static int find_sequence(int pid)
+{
+	char c, expectc = 'x';
+	int ycount = 0;
+	char dumpname[256];
+
+	snprintf(dumpname, 256, "dump-%d", pid);
+	tst_res(TINFO, "Dump file should be %s", dumpname);
+	if (access(dumpname, F_OK))
+		tst_brk(TCONF | TERRNO,
+			"Dump file was not found; assuming this is because "
+			"core dumps are disabled and not a bug");
+	tc.dfd = SAFE_OPEN(dumpname, O_RDONLY);
+
+	while (SAFE_READ(0, tc.dfd, &c, 1)) {
+		switch (c) {
+		case 'x':
+			ycount = 0;
+			expectc = 'y';
+			break;
+		case 'y':
+			if (expectc == 'y') {
+				ycount++;
+			} else {
+				expectc = 'x';
+				break;
+			}
+
+			if (ycount == tc.ycount)
+				expectc = 'z';
+			break;
+		case 'z':
+			if (expectc == 'z')
+				return 1;
+		default:
+			expectc = 'x';
+		}
+	}
+	return 0;
```

Comment:
This is a good idea. I wonder if we should randomize the data more, but that would require much more complicated code for the search and I certainly do not want to complicate the testcase code with algorithm that builds a search automata.